homeassistant/components/vorwerk/vacuum.py

```diff
@@ -0,0 +1,348 @@
+
+
+async def async_setup_entry(hass, entry, async_add_entities):
+    """Set up Vorwerk vacuum with config entry."""
+
+    _LOGGER.debug("Adding vorwerk vacuums")
+    async_add_entities(
+        [
+            VorwerkConnectedVacuum(robot)
+            for robot in hass.data[VORWERK_DOMAIN][entry.entry_id][VORWERK_ROBOTS]
+        ],
+        True,
+    )
+
+    platform = entity_platform.current_platform.get()
+    assert platform is not None
+
+    platform.async_register_entity_service(
+        "custom_cleaning",
+        {
+            vol.Optional(ATTR_MODE, default=2): cv.positive_int,
+            vol.Optional(ATTR_NAVIGATION, default=1): cv.positive_int,
+            vol.Optional(ATTR_CATEGORY, default=4): cv.positive_int,
+            vol.Optional(ATTR_ZONE): cv.string,
+        },
+        "vorwerk_custom_cleaning",
+    )
+
+
+class VorwerkConnectedVacuum(StateVacuumEntity):
+    """Representation of a Vorwerk Connected Vacuum."""
+
+    def __init__(self, robot):
+        """Initialize the Vorwerk Connected Vacuum."""
+        self.robot = robot
+        self._available = False
+        self._name = f"{self.robot.name}"
+        self._robot_has_map = False
+        self._robot_serial = self.robot.serial
+        self._status_state = None
+        self._clean_state = None
+        self._state = None
+        self._clean_time_start = None
+        self._clean_time_stop = None
+        self._clean_area = None
+        self._clean_battery_start = None
+        self._clean_battery_end = None
+        self._clean_susp_charge_count = None
+        self._clean_susp_time = None
+        self._clean_pause_time = None
+        self._clean_error_time = None
+        self._launched_from = None
+        self._battery_level = None
+        self._robot_boundaries = []
+        self._robot_stats = None
+
+    def update(self):
+        """Update the states of Vorwerk Vacuums."""
+        _LOGGER.debug("Running Vorwerk Vacuums update for '%s'", self.entity_id)
+        try:
+            if self._robot_stats is None:
+                self._robot_stats = self.robot.get_general_info().json().get("data")
+        except NeatoRobotException:
+            _LOGGER.warning("Couldn't fetch robot information of %s", self.entity_id)
+
+        try:
+            self._state = self.robot.state
+        except NeatoRobotException as ex:
+            if self._available:  # print only once when available
+                _LOGGER.error(
+                    "Vorwerk vacuum connection error for '%s': %s", self.entity_id, ex
+                )
+            self._state = None
+            self._available = False
+            return
+
+        self._available = True
+        _LOGGER.debug("self._state=%s", self._state)
+        if "alert" in self._state:
+            robot_alert = ALERTS.get(self._state["alert"])
+        else:
+            robot_alert = None
+        if self._state["state"] == 1:
+            if self._state["details"]["isCharging"]:
```

Comment:
why not use the ACTIONS and other const lists to look up the _clean_state and _status_state and fill them directly?

Reply:
I don't understand what's your idea here. Can you give me an example?

Reply:
> 
> 
> why not use the ACTIONS and other const lists to look up the _clean_state and _status_state and fill them directly?

I'm new to home assistant integrations could you show me an example?

Reply:
Sorry, I don't get it? Can you give some more hints?

Most of vacuum.py is a copy of the neato integration vacuum.py. For maintenance I would like to keep both version as near as possible.

Reply:
I do something similar in my integration: https://github.com/home-assistant/core/blob/fd68a3d23120214e6555ccc4f4af5185c53d70cb/homeassistant/components/sia/alarm_control_panel.py#L184, just getting the new_state from the big dict with conditions and actions and then executing it.

Reply:
I tried it today. It's not really getting better. The calculation of line 151-165 couldn't be easy mapped to a dict, because every STATE_CODE needs some thing different out of the state object.
When integrating the switch and sensor with the new dataupdatecoordinator, I will try to extract this calcuation.